src/mlpack/methods/reinforcement_learning/q_learning_impl.hpp

```diff
@@ -0,0 +1,204 @@
+             size_t explorationsSteps,
+             bool doubleQLearning,
+             size_t stepLimit,
+             EnvironmentType environment):
+    learningNetwork(network),
+    optimizer(optimizer),
+    discount(discount),
+    policy(std::move(policy)),
+    replayMethod(std::move(replayMethod)),
+    targetNetworkSyncInterval(targetNetworkSyncInterval),
+    explorationSteps(explorationsSteps),
+    doubleQLearning(doubleQLearning),
+    stepLimit(stepLimit),
+    environment(std::move(environment)),
+    totalSteps(0),
+    deterministic(false)
+{
+  learningNetwork.ResetParameters();
+  targetNetwork = learningNetwork;
+}
+
+
+template <
+  typename EnvironmentType,
+  typename NetworkType,
+  typename OptimizerType,
+  typename PolicyType,
+  typename ReplayType
+>
+arma::icolvec QLearning<
+  EnvironmentType,
+  NetworkType,
+  OptimizerType,
+  PolicyType,
+  ReplayType
+>::BestAction(const arma::mat& actionValues)
+{
+  arma::icolvec bestActions(actionValues.n_cols);
+  arma::rowvec maxActionValues = arma::max(actionValues, 0);
+  for (size_t i = 0; i < actionValues.n_cols; ++i)
+    bestActions(i) = arma::as_scalar(
+        arma::find(actionValues.col(i) == maxActionValues[i], 1));
+  return bestActions;
+};
+
+template <
+  typename EnvironmentType,
+  typename NetworkType,
+  typename OptimizerType,
+  typename BehaviorPolicyType,
+  typename ReplayType
+>
+double QLearning<
+  EnvironmentType,
+  NetworkType,
+  OptimizerType,
+  BehaviorPolicyType,
+  ReplayType
+>::Step()
+{
+  // Get the action value for each action at current state.
+  arma::colvec actionValue = learningNetwork.Predict(state.Encode());
+
+  // Select an action according to the behavior policy.
+  ActionType action = policy.Sample(actionValue, deterministic);
+
+  // Interact with the environment to advance to next state.
+  StateType nextState;
+  double reward = environment.Sample(state, action, nextState);
+
+  // Store the transition for replay.
+  replayMethod.Store(state, action, reward, nextState, environment.IsTerminal(nextState));
+
+  // Update current state.
+  state = nextState;
+
+  if (deterministic || totalSteps < explorationSteps)
+    return reward;
+
+  // Start experience replay.
+
+  // Sample from previous experience.
+  arma::mat sampledStates;
+  arma::icolvec sampledActions;
+  arma::colvec sampledRewards;
+  arma::mat sampledNextStates;
+  arma::icolvec isTerminal;
+  replayMethod.Sample(sampledStates, sampledActions, sampledRewards,
+                      sampledNextStates, isTerminal);
+
+  // Compute action value for next state with target network.
+  arma::mat nextActionValues = targetNetwork.Predict(sampledNextStates);
+
+  arma::icolvec bestActions;
+  if (doubleQLearning)
+    // If use double Q-Learning, use learning network to select the best action.
+    bestActions = BestAction(learningNetwork.Predict(sampledNextStates));
+  else
+    bestActions = BestAction(nextActionValues);
+
+  // Compute the update target.
```

Comment:
Sounds fine for me, do you like to do it in the context of this PR or later? I think if we expose the the ``Forward`` and ``Backward`` function for the ``FFN`` class we should do the same for the ``RNN`` class, just to be consistent.